### PR TITLE
Fix issue 123: Repeated reserialization of DAGs

### PR DESCRIPTION
**Pull request by rafidka**
*Issue #, if available:* #124

## Description of changes

Issue #124 happens because we are always calling `airflow db migrate` to make sure the database is migrated. The assumption was that the operation is a no-op if the DB is already initialized. However, this is not the case, and a reserialization of DAGs is happening every time this command is called. To avoid this, I am changing the code for executing migration to check if migration is needed.

## Testing

I did multiple tests using the Docker Compose setup. More testing by Amazon MWAA developers before this PR is merged. Suggested testing:

1. Create an internal 2.9.2 environment and make sure it is functioning normally. Also check ECS logs to ensure there are no issues.
2. Create a pre-2.9.2 environment and do an in-place upgrade to 2.9.2 and make sure the process is flawless.
3. Run a 2.9.2 environment with auto scaling enabled (ensuring that workers are being created and terminated continuously) and make sure migrations are not causing any issues.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [inline images/airflow/2.9.2/python/mwaa/database/migrate.py:54]: I'm not seeing `migratedb` in the docs, only [`migrate`](https://airflow.apache.org/docs/apache-airflow/2.9.2/cli-and-env-variables-ref.html#migrate)

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [inline images/airflow/2.9.2/python/mwaa/database/migrate.py:40]: Is a one second check sufficient or will we have false Timeout errors?  

Is the migrate command is idempotent to handle the case of false Timeout errors?

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [inline images/airflow/2.9.2/python/mwaa/database/migrate.py:41]: Just to confirm, this command is meant to wait until the migration is done?  So if the database is on a prior version we'd expect a quick timeout, otherwise it has a response we ignore.

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:40]: If always calling migrate was causing the re-serialization issue, it would seem this is not idempotent.  Is that accurate?

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:54]: Nevermind, I guess it's this:
https://github.com/apache/airflow/blob/36b92345b91e777f64dd860a740241a89c1232fc/airflow/cli/commands/db_command.py#L98

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (rafidka) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:40]: > Is a one second check sufficient or will we have false Timeout errors?

I need a method for checking whether the DB is initialized or not, but Airflow CLI didn't provide that unfortunately, so I had to employ the workaround of using `check_migrations` without a timeout.

> Is the migrate command is idempotent to handle the case of false Timeout errors?

Like above, I am using the check_migrations command to test whether migrations have been done or not, without a timeout.

> If always calling migrate was causing the re-serialization issue, it would seem this is not idempotent. Is that accurate?

Depends on what you mean by "this". Airflow's `db migrate` command is certainly not idempotent. However, the `_migrate_db` command is idempotent, and the reason is that we have it surrounded by `with_db_lock` meaning that no two processes can execute it at the same time, and within each call, we check whether the DB is initialized or not, and only initialize it if it is not.

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (rafidka) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:41]: > Just to confirm, this command is meant to wait until the migration is done?

Nope, it is just used as a method to check whether migrations have been applied or not.  See the detailed discussion above. I think I should refactor the code a bit to make it clear we are using check_migrations differently.

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (rafidka) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:54]: Exactly. The name in code doesn't match the name in the CLI.

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (dhegberg) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:40]: So with 2 (or more) schedulers.  

Previously one would get the lock, run migrate and release the lock.  The other would wait until the lock was released and run migrate resulting in the re-serialization.

Now one would get the lock, check the migration, run migrate and release the lock.  The other would wait until the lock was release and check the migration and exit.

Makes sense.

**Review by dhegberg — APPROVED** [review]

File: images/airflow/2.9.2/python/mwaa/database/migrate.py
Comment (ms964107) [reply images/airflow/2.9.2/python/mwaa/database/migrate.py:41]: Ref: https://github.com/apache/airflow/blob/36b92345b91e777f64dd860a740241a89c1232fc/airflow/utils/db.py#L840

**Review by ms964107 — APPROVED** [review]

**Comment by jaklan** [comment]
When it's merged, will MWAA 2.9.2-based instances be automatically updated? We are also affected by https://github.com/apache/airflow/issues/40082 and waiting for the fix before we propagate upgrade from 2.8 to 2.9 on PRD

**Comment by rafidka** [comment]
> When it's merged, will MWAA 2.9.2-based instances be automatically updated? We are also affected on DEV & UAT environments by [apache/airflow#40082](https://github.com/apache/airflow/issues/40082) and waiting for the fix before we propagate upgrade from 2.8 to 2.9 on PRD

Unfortunately no, you would need to trigger an environment update to have the Docker containers updated. Even a no-op update would do, though, so it should be pretty easy to do.

**Comment by jaklan** [comment]
@rafidka sure, that's not an issue. Are changes immediately propagated to MWAA after merging PR or is there any build- / deployment-related delay we have to take into consideration before we can trigger an update and test new containers?

**Comment by rafidka** [comment]
> Are changes immediately propagated to MWAA after merging PR or is there any build- / deployment-related delay we have to take into consideration before we can trigger an update and test new containers?

After the PR is merged, we sync it internally and deploy it via internal pipelines where it goes through testing, and then becomes available to different regions around the world gradually. Usually, assuming no test failures or pipeline blockages, this takes a couple of days. I am currently out of office, but @dhegberg should be able to help you with status updates. Daniel, we need to have this PR merged asap, as it is a production issue for our customers.

**Review by Mercury2699 — APPROVED** [review]

**Comment by jaklan** [comment]
MR is merged, so asking for the status update then 😄 Can we proceed with the re-deployment or nor yet?

**Comment by shubham22** [comment]
@Mercury2699 - can you confirm if this has been deployed to all the waves/regions?

**Comment by Mercury2699** [comment]
@jaklan Yes, this fix has been deployed to all regions. Customers can trigger environment update to receive the latest image.

**Pull request merged**
